ksql-common/src/main/java/io/confluent/ksql/function/UdfIndex.java

```diff
@@ -0,0 +1,317 @@
+        candidates.add(current);
+      }
+      return;
+    }
+
+    final Schema arg = arguments.get(argIndex);
+    current.children
+        .entrySet()
+        .stream()
+        .filter(entry -> entry.getKey().accepts(arg))
+        .map(Entry::getValue)
+        .forEach(node -> getCandidates(arguments, argIndex + 1, node, candidates));
+  }
+
+  private KsqlException createNoMatchingFunctionException(final List<Schema> paramTypes) {
+    final String sqlParamTypes = paramTypes.stream()
+        .map(schema -> schema == null
+            ? null
+            : SchemaUtil.getSchemaTypeAsSqlType(schema.type()))
+        .collect(Collectors.joining(", ", "[", "]"));
+
+    return new KsqlException("Function '" + metadata.getName()
+        + "' does not accept parameters of types:" + sqlParamTypes);
+  }
+
+  public Collection<KsqlFunction> values() {
+    return allFunctions.values();
+  }
+
+  @SuppressWarnings("unused")
+  @VisibleForTesting
+  String describe() {
+    final StringBuilder sb = new StringBuilder();
+    sb.append("-ROOT\n");
+    root.describe(sb, 1);
+    return sb.toString();
+  }
+
+  private static final class Node {
+
+    private static final Comparator<KsqlFunction> COMPARE_FUNCTIONS =
+        Comparator.nullsFirst(
+            Comparator
+                .<KsqlFunction, Integer>comparing(fun -> fun.isVarArgs() ? 0 : 1)
+                .thenComparing(fun -> fun.getArguments().size())
+        );
+
+    private final Map<Parameter, Node> children;
+    private KsqlFunction value;
+    private int order = 0;
+
+    private Node() {
+      this.children = new HashMap<>();
+      this.value = null;
+    }
+
+    private void update(final KsqlFunction function, final int order) {
+      if (COMPARE_FUNCTIONS.compare(function, value) > 0) {
+        value = function;
+        this.order = order;
+      }
+    }
+
+    private void describe(final StringBuilder builder, final int indent) {
+      for (final Entry<Parameter, Node> child : children.entrySet()) {
+        if (child.getValue() != this) {
+          builder.append(StringUtils.repeat(' ', indent * 2))
+              .append('-')
+              .append(child.getKey())
+              .append(": ")
+              .append(child.getValue())
+              .append('\n');
+          child.getValue().describe(builder, indent + 1);
+        }
+      }
+    }
+
+    @Override
+    public String toString() {
+      return value != null ? value.getFunctionName() : "EMPTY";
+    }
+
+    public static int compare(final Node a, final Node b) {
+      final int compare = COMPARE_FUNCTIONS.compare(a.value, b.value);
+      return compare == 0 ? -(a.order - b.order) : compare;
+    }
+  }
+
+  /**
+   * A class that represents a parameter, with a schema and whether
+   * or not it is part of a variable argument declaration.
+   */
+  private static final class Parameter {
+
+    private static final Map<Type, BiPredicate<Schema, Schema>> CUSTOM_SCHEMA_EQ =
```

Comment:
As it stands, the package setup wouldn't let me use `PrimitiveType`... since this code is in ksql-common which doesn't depend on ksql-parser... I will go with the full list

Reply:
Actually now that I think of it, a full list doesn't really help. The reason I'm adding this instead of always checking these components is that calling `keySchema/valueSchema/fields` on primitives throws an exception. Since this is a type map, not a schema map, the list is pretty much set in stone. For the example you gave the workaround would need to be at the schema level not the type level (and actually, I think it would work out of the box since the decimal is of type struct from what i understand).

Reply:
In the spirit of incremental progress, I will revisit this when I introduce the generic/struct support for UDFs